File: src/main/java/com/qwazr/jdbc/cache/Driver.java
/*
 * Copyright 2016 Emmanuel Keller / QWAZR
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.qwazr.jdbc.cache;

import java.nio.file.FileSystems;
import java.nio.file.Path;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.DriverPropertyInfo;
import java.sql.SQLException;
import java.sql.SQLFeatureNotSupportedException;
import java.util.Properties;
import java.util.concurrent.ConcurrentHashMap;
import java.util.logging.Level;
import java.util.logging.Logger;

public class Driver implements java.sql.Driver {

    final static Logger LOGGER = Logger.getLogger(Driver.class.getPackage().getName());

    public final static String URL_FILE_PREFIX = "jdbc:cache:file:";
    public final static String URL_MEM_PREFIX = "jdbc:cache:mem:";
    public final static String CACHE_DRIVER_URL = "cache.driver.url";
    public final static String CACHE_DRIVER_CLASS = "cache.driver.class";
    public final static String CACHE_DRIVER_ACTIVE = "cache.driver.active";

    static {
        try {
            DriverManager.registerDriver(new Driver());
        } catch (SQLException e) {
            LOGGER.log(Level.SEVERE, e.getMessage(), e);
        }
    }

    private final ConcurrentHashMap<String, ResultSetCache> resultSetCacheMap = new ConcurrentHashMap<>();

    public Connection connect(String url, Properties info) throws SQLException {

        if (!acceptsURL(url))
            return null;

        // Determine the optional backend connection
        final String cacheDriverUrl = info.getProperty(CACHE_DRIVER_URL);
        final String cacheDriverClass = info.getProperty(CACHE_DRIVER_CLASS);
        try {
            if (cacheDriverClass != null && !cacheDriverClass.isEmpty())
                Class.forName(cacheDriverClass);
        } catch (ClassNotFoundException e) {
            throw new SQLException("Cannot initialize the driver: " + cacheDriverClass, e);
        }

        final String cacheDriverActive = info.getProperty(CACHE_DRIVER_ACTIVE);
        final boolean active = cacheDriverActive == null || Boolean.parseBoolean(cacheDriverActive);

        final Connection backendConnection = cacheDriverUrl == null || cacheDriverUrl.isEmpty() ?
                null :
                DriverManager.getConnection(cacheDriverUrl, info);

        if (!active) {
            return new CachedConnection(backendConnection, null);
        }

        final ResultSetCache resultSetCache;
        if (url.startsWith(URL_FILE_PREFIX)) {
            if (url.length() <= URL_FILE_PREFIX.length()) {
                throw new SQLException("The path is empty: " + url);
            }
            // Check the cache directory
            final String cacheName = url.substring(URL_FILE_PREFIX.length());
            final Path cacheDirectory = FileSystems.getDefault().getPath(cacheName);
            resultSetCache = resultSetCacheMap.computeIfAbsent(cacheName, (foo) -> new ResultSetOnDiskCacheImpl(cacheDirectory));
        } else if (url.startsWith(URL_MEM_PREFIX)) {
            if (url.length() <= URL_MEM_PREFIX.length()) {
                throw new SQLException("The name is empty: " + url);
            }
            // Check the cache directory
            final String cacheName = url.substring(URL_MEM_PREFIX.length());
            resultSetCache = resultSetCacheMap.computeIfAbsent(cacheName, (foo) -> new ResultSetInMemoryCacheImpl());
        } else {
            throw new IllegalArgumentException("Can not find cache implementation for " + url);
        }

        return new CachedConnection(backendConnection, resultSetCache);
    }

    public boolean acceptsURL(String url) throws SQLException {
        return url != null && (url.startsWith(URL_FILE_PREFIX) || url.startsWith(URL_MEM_PREFIX));
    }

    public DriverPropertyInfo[] getPropertyInfo(final String url, final Properties info) throws SQLException {
        return new DriverPropertyInfo[] { new DriverPropertyInfo(CACHE_DRIVER_URL, null),
                new DriverPropertyInfo(CACHE_DRIVER_CLASS, null), new DriverPropertyInfo(CACHE_DRIVER_ACTIVE, null) };
    }

    public int getMajorVersion() {
        return 1;
    }

    public int getMinorVersion() {
        return 3;
    }

    public boolean jdbcCompliant() {
        return false;
    }

    public Logger getParentLogger() throws SQLFeatureNotSupportedException {
        return LOGGER;
    }

    public static ResultSetCache getCache(final Connection connection) throws SQLException {
        if (!(connection instanceof CachedConnection))
            throw new SQLException("The connection is not a cached connection");
        return ((CachedConnection) connection).getResultSetCache();
    }

}

File: src/main/java/com/statewidesoftware/jdbc/cache/Driver.kt
/*
 * Copyright 2016 Emmanuel Keller / QWAZR
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.statewidesoftware.jdbc.cache

import java.nio.file.FileSystems
import java.sql.*
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.logging.Level
import java.util.logging.Logger

class Driver : java.sql.Driver {
    private val resultSetCacheMap = ConcurrentHashMap<String, ResultSetCache>()
    @Throws(SQLException::class, IllegalArgumentException::class)
    override fun connect(url: String, info: Properties): Connection? {
        if (!acceptsURL(url)) return null

        // Determine the optional backend connection
        val cacheDriverUrl = info.getProperty(CACHE_DRIVER_URL)
        val cacheDriverClass = info.getProperty(CACHE_DRIVER_CLASS)
        try {
            if (cacheDriverClass != null && cacheDriverClass.isNotEmpty()) Class.forName(cacheDriverClass)
        } catch (e: ClassNotFoundException) {
            throw SQLException("Cannot initialize the driver: $cacheDriverClass", e)
        }
        val cacheDriverActive = info.getProperty(CACHE_DRIVER_ACTIVE)
        val active = cacheDriverActive == null || java.lang.Boolean.parseBoolean(cacheDriverActive)
        val backendConnection =
            if (cacheDriverUrl == null || cacheDriverUrl.isEmpty()) null else DriverManager.getConnection(
                cacheDriverUrl,
                info
            )
        if (!active) {
            return CachedConnection(backendConnection, null)
        }
        val resultSetCache: ResultSetCache = if (url.startsWith(URL_FILE_PREFIX)) {
            if (url.length <= URL_FILE_PREFIX.length) {
                throw SQLException("The path is empty: $url")
            }
            // Check the cache directory
            val cacheName = url.substring(URL_FILE_PREFIX.length)
            val cacheDirectory = FileSystems.getDefault().getPath(cacheName)
            resultSetCacheMap.computeIfAbsent(cacheName) { ResultSetOnDiskCacheImpl(cacheDirectory) }
        } else if (url.startsWith(URL_MEM_PREFIX)) {
            if (url.length <= URL_MEM_PREFIX.length) {
                throw SQLException("The name is empty: $url")
            }
            // Check the cache directory
            val cacheName = url.substring(URL_MEM_PREFIX.length)
            resultSetCacheMap.computeIfAbsent(cacheName) { ResultSetInMemoryCacheImpl() }
        } else {
            throw IllegalArgumentException("Can not find cache implementation for $url")
        }
        return CachedConnection(backendConnection, resultSetCache)
    }

    @Throws(SQLException::class)
    override fun acceptsURL(url: String?): Boolean {
        return (url != null) && (url.startsWith(URL_FILE_PREFIX) || url.startsWith(URL_MEM_PREFIX))
    }

    @Throws(SQLException::class)
    override fun getPropertyInfo(url: String, info: Properties): Array<DriverPropertyInfo> {
        return arrayOf(
            DriverPropertyInfo(CACHE_DRIVER_URL, null),
            DriverPropertyInfo(CACHE_DRIVER_CLASS, null), DriverPropertyInfo(CACHE_DRIVER_ACTIVE, null)
        )
    }

    override fun getMajorVersion(): Int {
        return 1
    }

    override fun getMinorVersion(): Int {
        return 3
    }

    override fun jdbcCompliant(): Boolean {
        return false
    }

    @Throws(SQLFeatureNotSupportedException::class)
    override fun getParentLogger(): Logger {
        return LOGGER
    }

    companion object {
        val LOGGER = Logger.getLogger(Driver::class.java.getPackage().name)
        const val URL_FILE_PREFIX = "jdbc:cache:file:"
        const val URL_MEM_PREFIX = "jdbc:cache:mem:"
        const val CACHE_DRIVER_URL = "cache.driver.url"
        const val CACHE_DRIVER_CLASS = "cache.driver.class"
        const val CACHE_DRIVER_ACTIVE = "cache.driver.active"

        init {
            try {
                DriverManager.registerDriver(Driver())
            } catch (e: SQLException) {
                LOGGER.log(Level.SEVERE, e.message, e)
            }
        }

        @JvmStatic
        @Throws(SQLException::class)
        fun getCache(connection: Connection?): ResultSetCache? {
            if (connection !is CachedConnection) throw SQLException("The connection is not a cached connection")
            return connection.resultSetCache
        }
    }
}